src/core/mesh/qgsmeshtracerenderer.cpp

```diff
@@ -0,0 +1,605 @@
+  int fieldLeftDevice = mFieldTopLeftInDeviceCoordinates.y();
+  QgsPointXY fieldTopLeftMap = deviceMapToPixel.toMapCoordinates( fieldTopDevice, fieldLeftDevice );
+
+  double xc = ( fieldRightBottomMap.x() + fieldTopLeftMap.x() ) / 2;
+  double yc = ( fieldTopLeftMap.y() + fieldRightBottomMap.y() ) / 2;
+
+  mMapToFieldPixel = QgsMapToPixel( mapUnitperFieldPixel,
+                                    xc,
+                                    yc,
+                                    fieldWidth,
+                                    fieldHeight, 0 );
+
+  initField();
+  mValid = true;
+}
+
+void QgsMeshTraceField::addTrace( QgsPointXY startPoint )
+{
+  addTrace( mMapToFieldPixel.transform( startPoint ).toQPointF().toPoint() );
+}
+
+void QgsMeshTraceField::addRandomTraces( int count )
+{
+  for ( int i = 0; i < count; ++i )
+    addRandomTrace();
+}
+
+void QgsMeshTraceField::addRandomTrace()
+{
+  if ( !mValid )
+    return;
+
+  int xRandom =  1 + std::rand() / int( ( RAND_MAX + 1u ) / uint( mFieldSize.width() ) )  ;
+  int yRandom = 1 + std::rand() / int ( ( RAND_MAX + 1u ) / uint( mFieldSize.height() ) ) ;
+
+  addTrace( QPoint( xRandom, yRandom ) );
+}
+
+void QgsMeshTraceField::addGriddedTraces( int pixelSpace )
+{
+  int fieldSpace = pixelSpace / mFieldResolution;
+  int i = 0;
+  while ( i < size().width() )
+  {
+    int j = 0;
+    while ( j < size().height() )
+    {
+      addTrace( QPoint( i, j ) );
+      j += fieldSpace;
+    }
+    i += fieldSpace;
+  }
+}
+
+void QgsMeshTraceField::addTracesFromBorder( int pixelSpace )
+{
+  int fieldSpace = pixelSpace / mFieldResolution;
+  int i = 0;
+  int j = 0;
+  while ( j < size().height() )
+  {
+    addTrace( QPoint( i, j ) );
+    j += fieldSpace;
+  }
+  i = size().width() - 1;
+  j = 0;
+  while ( j < size().height() )
+  {
+    addTrace( QPoint( i, j ) );
+    j += fieldSpace;
+  }
+  i = 0;
+  j = 0;
+  while ( i < size().width() )
+  {
+    addTrace( QPoint( i, j ) );
+    i += fieldSpace;
+  }
+  i = 0;
+  j = size().height() - 1;
+  while ( i < size().width() )
+  {
+    addTrace( QPoint( i, j ) );
+    i += fieldSpace;
+  }
+
+}
+
+void QgsMeshTraceField::addTrace( QPoint startPixel )
+{
+  //This is where each trace are contructed
+
+  if ( !mValid )
+    return;
+
+  if ( isWayExist( startPixel ) )
+    return;
+
+  if ( ! mVectorValueInterpolator )
+    return;
+
+  bool end = false;
+  //position int the pixelField
+  double x1 = 0;
+  double y1 = 0;
+
+  QPoint currentPixel = startPixel;
+  QgsVector vector;
+  startTrace( currentPixel );
+
+  float dt = 0;
+
+  int i = 0;
+  while ( !end )
+  {
+    QgsPointXY mapPosition = positionToMapCoordinnate( currentPixel, QgsPointXY( x1, y1 ) );
+    vector = mVectorValueInterpolator->value( mapPosition ) ;
+
+    if ( std::isnan( vector.x() ) || std::isnan( vector.y() ) )
+    {
+      endTrace( currentPixel );
+      break;
+    }
+
+    /* adimentional value :  Vu=2 when the particule need dt=1 to go throught a pixel
```

Comment:
typo: "nondimensional" (adimentional is spanish?)

Reply:
no frenglish... :)